vendor/bat-native-ledger/src/bat/ledger/internal/publisher/publisher_list_updater.cc

```diff
@@ -0,0 +1,159 @@
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "bat/ledger/internal/publisher/publisher_list_updater.h"
+
+#include <memory>
+#include <utility>
+
+#include "bat/ledger/internal/common/time_util.h"
+#include "bat/ledger/internal/ledger_impl.h"
+#include "bat/ledger/internal/publisher/publisher_list_reader.h"
+#include "bat/ledger/internal/request/request_publisher.h"
+#include "bat/ledger/internal/state/state_keys.h"
+#include "bat/ledger/option_keys.h"
+#include "net/http/http_status_code.h"
+
+using std::placeholders::_1;
+using std::placeholders::_2;
+using std::placeholders::_3;
+
+namespace {
+
+constexpr int64_t kRetryAfterFailureDelay = 150;
+constexpr int64_t kMaxRetryAfterFailureDelay = 4 * base::Time::kSecondsPerHour;
+
+}  // namespace
+
+namespace braveledger_publisher {
+
+PublisherListUpdater::PublisherListUpdater(bat_ledger::LedgerImpl* ledger)
+    : ledger_(ledger) {}
+
+PublisherListUpdater::~PublisherListUpdater() = default;
+
+void PublisherListUpdater::StartAutoUpdate() {
+  auto_update_ = true;
+  if (!timer_.IsRunning()) {
+    StartFetchTimer(FROM_HERE, GetAutoUpdateDelay());
+  }
+}
+
+void PublisherListUpdater::StopAutoUpdate() {
+  BLOG(1, "Cancelling publisher prefix list update");
+  auto_update_ = false;
+  timer_.Stop();
+}
+
+void PublisherListUpdater::SetOnPublisherListUpdatedCallback(
+    OnPublisherListUpdatedCallback callback) {
+  on_updated_callback_ = callback;
+}
+
+void PublisherListUpdater::StartFetchTimer(
+    const base::Location& posted_from,
+    base::TimeDelta delay) {
+  BLOG(1, "Scheduling publisher prefix list update in "
+      << delay.InSeconds() << " seconds");
+  timer_.Start(posted_from, delay, base::BindOnce(
+      &PublisherListUpdater::OnFetchTimerElapsed,
+      base::Unretained(this)));
+}
+
+void PublisherListUpdater::OnFetchTimerElapsed() {
+  BLOG(1, "Fetching publisher prefix list");
+  std::string url = braveledger_request_util::GetPublisherListUrl();
+  ledger_->LoadURL(
+      url, {}, "", "",
+      ledger::UrlMethod::GET,
+      std::bind(&PublisherListUpdater::OnFetchCompleted, this, _1));
+}
+
+void PublisherListUpdater::OnFetchCompleted(
+    const ledger::UrlResponse& response) {
+  if (response.status_code != net::HTTP_OK || response.body.empty()) {
+    BLOG(0, "Invalid server response for publisher prefix list");
+    StartFetchTimer(FROM_HERE, GetRetryAfterFailureDelay());
+    return;
+  }
+
+  auto reader = std::make_unique<PublisherListReader>();
+  auto parse_error = reader->Parse(response.body);
+  if (parse_error != PublisherListReader::ParseError::None) {
+    // This could be a problem on the client or the server, but
+    // optimistically assume that it is a server issue and retry
+    // with back-off.
+    BLOG(0, "Failed to parse publisher prefix list: "
+        << static_cast<int>(parse_error));
```

Comment:
cast is not needed

Reply:
The cast is needed since `ParseError` is an enum class.